flit_core/flit_core/config.py

```diff
@@ -822,34 +824,96 @@ def isabs_ish(path):
     return os.path.isabs(path) or path.startswith(('/', '\\'))
 
 
-def normalize_license_expr(s: str):
-    """Validate & normalise an SPDX license expression
+def normalise_compound_license_expr(s: str) -> str:
+    """Validate and normalise a compund SPDX license expression.
 
-    For now this only handles simple expressions (referring to 1 license)
+    Per the specification, licence expression operators (AND, OR and WITH)
+    are matched case-sensitively. The WITH operator is not currently supported.
+
+    Spec: https://spdx.github.io/spdx-spec/v2.2.2/SPDX-license-expressions/
+    """
+    invalid_msg = "'{s}' is not a valid SPDX license expression: {reason}"
+    if not s or s.isspace():
+        raise ConfigError(f"The SPDX license expression must not be empty")
+
+    stack = 0
+    parts = []
+    try:
+        for part in filter(None, re.split(r' +|([()])', s)):
+            if part.upper() == 'WITH':
+                # provide a sensible error message for the WITH operator
+                raise ConfigError(f"The SPDX 'WITH' operator is not yet supported!")
+            elif part in {'AND', 'OR'}:
+                if not parts or parts[-1] in {' AND ', ' OR ', ' WITH ', '('}:
+                    reason = f"'{part}' must follow a license ID"
+                    raise ConfigError(invalid_msg.format(s=s, reason=reason))
+                parts.append(f' {part} ')
+            elif part.lower() in {'and', 'or', 'with'}:
+                # provide a sensible error message for non-uppercase operators
+                reason = f"operators must be uppercase, not '{part}'"
+                raise ConfigError(invalid_msg.format(s=s, reason=reason))
+            elif part == '(':
+                if parts and parts[-1] not in {' AND ', ' OR ', '('}:
+                    reason = f"'(' must follow either AND, OR, or '('"
+                    raise ConfigError(invalid_msg.format(s=s, reason=reason))
+                stack += 1
+                parts.append(part)
+            elif part == ')':
+                if not parts or parts[-1] in {' AND ', ' OR ', ' WITH ', '('}:
+                    reason = "')' must follow a license ID"
```

Comment:
```py
'(MIT AND)' is not a valid SPDX license expression: ')' must follow a license ID
```
Maybe the reason could be improved here. It's not so much about the `)` position but rather that the second license id is missing.

Reply:
I've rephrased the error messages